modules/ingestor/src/service/advisory/csaf/creator.rs

```diff
@@ -0,0 +1,147 @@
+use crate::{
+    graph::{
+        advisory::advisory_vulnerability::{VersionInfo, VersionSpec},
+        purl::creator::PurlCreator,
+    },
+    service::{advisory::csaf::util::resolve_purls, Error},
+};
+use csaf::{definitions::ProductIdT, Csaf};
+use sea_orm::{ActiveValue::Set, ColumnTrait, ConnectionTrait, EntityTrait, QueryFilter};
+use sea_query::IntoCondition;
+use std::{collections::hash_map::Entry, collections::HashMap};
+use trustify_common::{db::chunk::EntityChunkedIter, purl::Purl};
+use trustify_entity::{package_status, status, version_range};
+use uuid::Uuid;
+
+struct PackageStatus {
+    package: Purl,
+    status: &'static str,
+    info: VersionInfo,
+}
+
+pub struct PackageStatusCreator {
+    advisory_id: Uuid,
+    vulnerability_id: i32,
+    entries: Vec<PackageStatus>,
+}
+
+impl PackageStatusCreator {
+    pub fn new(advisory_id: Uuid, vulnerability_id: i32) -> Self {
+        Self {
+            advisory_id,
+            vulnerability_id,
+            entries: Vec::new(),
+        }
+    }
+
+    pub fn add_all(&mut self, csaf: &Csaf, ps: &Option<Vec<ProductIdT>>, status: &'static str) {
+        for r in ps.iter().flatten() {
+            for purl in resolve_purls(csaf, r) {
+                let package = Purl::from(purl.clone());
+
+                if let Some(version) = package.version.clone() {
+                    self.entries.push(PackageStatus {
+                        package,
+                        status,
+                        info: VersionInfo {
+                            scheme: "generic".to_string(),
+                            spec: VersionSpec::Exact(version),
+                        },
+                    });
+                }
+            }
+        }
+    }
+
+    async fn check_status(
+        status: &str,
+        connection: &impl ConnectionTrait,
+    ) -> Result<status::Model, Error> {
+        Ok(status::Entity::find()
+            .filter(status::Column::Slug.eq(status))
+            .one(connection)
+            .await?
+            .ok_or_else(|| crate::graph::error::Error::InvalidStatus(status.to_string()))?)
+    }
+
+    pub async fn create(self, connection: &impl ConnectionTrait) -> Result<(), Error> {
+        let mut checked = HashMap::new();
+
+        let mut purls = PurlCreator::new();
+
+        for ps in &self.entries {
+            // ensure a correct status, and get id
+            if let Entry::Vacant(entry) = checked.entry(ps.status) {
+                entry.insert(Self::check_status(ps.status, connection).await?);
```

Comment:
Does this mean status rows can be implicitly added upon first sighting?  I think that's fine.

Reply:
Nope, that just inserts it into the hash map. The idea is to look it up only once, cache the `status.id` and re-use it a bit further down the function. Versus: looking it up for each entry.